<a href="https://colab.research.google.com/github/olhaishchenko/-data_science/blob/master/H_W_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
pip install scikit-surprise

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 772.0/772.0 kB 5.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.3-cp310-cp310-linux_x86_64.whl size=2811631 sha256=71c7897b3b300778b8ade774854794f500fa237b104ac5f4443418ca1932b67b
  Stored in directory: /root/.cache/pip/wheels/a5/ca/a8/4e28def53797fdc4363ca4af740db15a9c2f1595ebc51fb445
Successfully built scikit-surprise


In [ ]:
import pandas as pd

from surprise import Dataset, SVD, NMF, SVDpp, KNNBasic
from collections import defaultdict
from surprise.model_selection import train_test_split, cross_validate, GridSearchCV

In [ ]:
data = Dataset.load_builtin("ml-100k")

In [ ]:
trainset, testset = train_test_split(data, test_size=0.05)#test_size=0.05??

In [ ]:
param_grid = {"n_epochs": [5, 10], "lr_all": [0.002, 0.005], "reg_all": [0.4, 0.6]}
gs = GridSearchCV(SVD, param_grid, measures=["rmse", "mae"], cv=3)

gs.fit(data)

# best RMSE score
print(gs.best_score["rmse"])

# combination of parameters that gave the best RMSE score
print(gs.best_params["rmse"])

0.9634763122313922
{'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.4}


In [ ]:
algo_svd = SVD(n_epochs=10, lr_all=0.005, reg_all=0.4, random_state=0)
# Run 5-fold cross-validation and print results.
print(f"cross_val_SVD")
cross_validate(algo_svd, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

cross_val_SVD
Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9600  0.9683  0.9525  0.9667  0.9589  0.9613  0.0057  
MAE (testset)     0.7695  0.7793  0.7594  0.7752  0.7691  0.7705  0.0067  
Fit time          0.70    0.87    0.69    0.99    0.94    0.84    0.12    
Test time         0.16    0.20    0.16    0.26    0.16    0.19    0.04    


{'test_rmse': array([0.96003595, 0.96829748, 0.95251132, 0.96674872, 0.95893905]),
 'test_mae': array([0.7694649 , 0.77933139, 0.75942296, 0.77522629, 0.76907737]),
 'fit_time': (0.7003955841064453,
  0.8724386692047119,
  0.6948111057281494,
  0.9919626712799072,
  0.9385859966278076),
 'test_time': (0.15776395797729492,
  0.19711661338806152,
  0.1637401580810547,
  0.25684547424316406,
  0.15615534782409668)}

In [ ]:
algo_knn = KNNBasic(k=200)
print(f"cross_val_KNN")
cross_validate(algo_knn, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

cross_val_KNN
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Evaluating RMSE, MAE of algorithm KNNBasic on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    1.0136  0.9995  1.0070  0.9942  0.9969  1.0023  0.0071  
MAE (testset)     0.8057  0.7930  0.7996  0.7901  0.7908  0.7959  0.0060  
Fit time          0.47    0.38    0.72    0.35    0.65    0.51    0.15    
Test time         4.58    5.24    4.66    4.54    5.38    4.88    0.36    


{'test_rmse': array([1.0136086 , 0.99954258, 1.007048  , 0.99418046, 0.99688217]),
 'test_mae': array([0.80574879, 0.79303753, 0.79963131, 0.79011902, 0.79084317]),
 'fit_time': (0.4729151725769043,
  0.3803389072418213,
  0.7216660976409912,
  0.34561872482299805,
  0.6537837982177734),
 'test_time': (4.579036235809326,
  5.240965843200684,
  4.659076690673828,
  4.537182807922363,
  5.375331401824951)}

In [ ]:
param_grid = {"n_epochs": [25, 40, 50], "biased": [True, False], "n_factors": [5, 10, 15]}
gs = GridSearchCV(NMF, param_grid, measures=["rmse", "mae"], cv=3)

gs.fit(data)

# best RMSE score
print(gs.best_score["rmse"])

# combination of parameters that gave the best RMSE score
print(gs.best_params["rmse"])

0.9404781008612515
{'n_epochs': 50, 'biased': True, 'n_factors': 5}


In [ ]:
algo_nmf = NMF(random_state=0)
print(f"cross_val_NMF")
cross_validate(algo_nmf, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

cross_val_NMF
Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9652  0.9543  0.9660  0.9566  0.9590  0.9602  0.0047  
MAE (testset)     0.7610  0.7503  0.7600  0.7511  0.7529  0.7551  0.0045  
Fit time          2.25    2.84    3.69    2.44    2.25    2.69    0.54    
Test time         0.19    0.17    0.14    0.20    0.13    0.17    0.03    


{'test_rmse': array([0.96524969, 0.95429236, 0.96598725, 0.95655108, 0.95903996]),
 'test_mae': array([0.76102528, 0.75034688, 0.75995745, 0.75110625, 0.7529019 ]),
 'fit_time': (2.2471985816955566,
  2.8438963890075684,
  3.692803382873535,
  2.444042921066284,
  2.2467780113220215),
 'test_time': (0.18850255012512207,
  0.1721055507659912,
  0.13794350624084473,
  0.20195960998535156,
  0.13450169563293457)}

In [ ]:
param_grid = {"n_epochs": [10, 20], "init_std_dev": [0.1, 0.05], "n_factors": [5, 10, 20]}
gs = GridSearchCV(SVDpp, param_grid, measures=["rmse", "mae"], cv=3)

gs.fit(data)

# best RMSE score
print(gs.best_score["rmse"])

# combination of parameters that gave the best RMSE score
print(gs.best_params["rmse"])

0.9188746681241698
{'n_epochs': 20, 'init_std_dev': 0.05, 'n_factors': 20}


In [ ]:
algo_svdpp = SVDpp(init_std_dev=0.05, random_state=0)
print(f"cross_val_SVDpp")
cross_validate(algo_svdpp, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

cross_val_SVDpp
Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9109  0.9061  0.9111  0.9121  0.9100  0.9100  0.0021  
MAE (testset)     0.7179  0.7135  0.7149  0.7148  0.7126  0.7147  0.0018  
Fit time          27.26   27.30   28.05   27.17   27.02   27.36   0.36    
Test time         5.68    4.57    5.68    6.05    4.78    5.35    0.57    


{'test_rmse': array([0.91089612, 0.90608158, 0.91112978, 0.91208725, 0.90996401]),
 'test_mae': array([0.7179118 , 0.71345397, 0.71490083, 0.71482557, 0.7126277 ]),
 'fit_time': (27.25819730758667,
  27.299097299575806,
  28.050058603286743,
  27.174739837646484,
  27.019202709197998),
 'test_time': (5.683303117752075,
  4.567030191421509,
  5.678992509841919,
  6.047089338302612,
  4.78488826751709)}

| [Movielens 100k](https://grouplens.org/datasets/movielens/100k)                                                                         |   RMSE |   MAE | Time    |

| [SVD](https://surprise.readthedocs.io/en/stable/matrix_factorization.html#surprise.prediction_algorithms.matrix_factorization.SVD)      |  0.934 | 0.737 | 0:00:09 |

| [SVD++](https://surprise.readthedocs.io/en/stable/matrix_factorization.html#surprise.prediction_algorithms.matrix_factorization.SVDpp)  |  0.919 | 0.721 | 0:02:46 |

| [SVD++](https://surprise.readthedocs.io/en/stable/matrix_factorization.html#surprise.prediction_algorithms.matrix_factorization.SVDpp)  |  0.919 | 0.721 | 0:02:20 |

| [NMF](https://surprise.readthedocs.io/en/stable/matrix_factorization.html#surprise.prediction_algorithms.matrix_factorization.NMF)      |  0.963 | 0.758 | 0:00:17 |

| [k-NN](https://surprise.readthedocs.io/en/stable/knn_inspired.html#surprise.prediction_algorithms.knns.KNNBasic)                        |  0.98  | 0.774 | 0:00:24 |

In [ ]:
algo_svd.fit(trainset)

In [ ]:
preds_svd = algo_svd.test(testset)
preds_svd[:10]

[Prediction(uid='416', iid='421', r_ui=5.0, est=3.918256745201451, details={'was_impossible': False}),
 Prediction(uid='327', iid='475', r_ui=4.0, est=3.5546980497898706, details={'was_impossible': False}),
 Prediction(uid='6', iid='496', r_ui=4.0, est=3.839293377278706, details={'was_impossible': False}),
 Prediction(uid='249', iid='300', r_ui=4.0, est=3.9843593762832588, details={'was_impossible': False}),
 Prediction(uid='6', iid='297', r_ui=3.0, est=3.750256778103611, details={'was_impossible': False}),
 Prediction(uid='189', iid='657', r_ui=5.0, est=4.247007163422771, details={'was_impossible': False}),
 Prediction(uid='663', iid='282', r_ui=3.0, est=3.6815938616384436, details={'was_impossible': False}),
 Prediction(uid='807', iid='1444', r_ui=3.0, est=3.2985414737955603, details={'was_impossible': False}),
 Prediction(uid='688', iid='749', r_ui=5.0, est=4.194749986071955, details={'was_impossible': False}),
 Prediction(uid='560', iid='240', r_ui=3.0, est=2.9010727160565994, deta

In [ ]:
algo_knn.fit(trainset)

Computing the msd similarity matrix...
Done computing similarity matrix.


In [ ]:
preds_knn = algo_knn.test(testset)
preds_knn[:10]

[Prediction(uid='416', iid='421', r_ui=5.0, est=3.68323553520574, details={'actual_k': 101, 'was_impossible': False}),
 Prediction(uid='327', iid='475', r_ui=4.0, est=3.8779325083890965, details={'actual_k': 200, 'was_impossible': False}),
 Prediction(uid='6', iid='496', r_ui=4.0, est=4.1601710236607685, details={'actual_k': 200, 'was_impossible': False}),
 Prediction(uid='249', iid='300', r_ui=4.0, est=3.7525481897114683, details={'actual_k': 200, 'was_impossible': False}),
 Prediction(uid='6', iid='297', r_ui=3.0, est=3.962322254302657, details={'actual_k': 48, 'was_impossible': False}),
 Prediction(uid='189', iid='657', r_ui=5.0, est=4.30337009899607, details={'actual_k': 129, 'was_impossible': False}),
 Prediction(uid='663', iid='282', r_ui=3.0, est=3.734018676196836, details={'actual_k': 200, 'was_impossible': False}),
 Prediction(uid='807', iid='1444', r_ui=3.0, est=2.609057689042965, details={'actual_k': 18, 'was_impossible': False}),
 Prediction(uid='688', iid='749', r_ui=5.0, 

In [ ]:
from surprise import accuracy

In [ ]:
accuracy.mse(preds_svd), accuracy.mae(preds_svd)#mae краще

MSE: 0.9196
MAE:  0.7663


(0.9196197464255691, 0.7662579578276479)

In [ ]:
accuracy.mse(preds_knn), accuracy.mae(preds_knn)

MSE: 0.9955
MAE:  0.7893


(0.9955464678691386, 0.789288148412357)

# **рекомендації для юзерів**

In [ ]:
trainset1, testset1 = train_test_split(data, test_size=0.001)

In [ ]:
predictions = algo_knn.test(trainset1.build_anti_testset())

In [ ]:
predictions[:5]

[Prediction(uid='405', iid='150', r_ui=3.52990990990991, est=3.889823903524877, details={'actual_k': 146, 'was_impossible': False}),
 Prediction(uid='405', iid='124', r_ui=3.52990990990991, est=3.917481332041734, details={'actual_k': 172, 'was_impossible': False}),
 Prediction(uid='405', iid='748', r_ui=3.52990990990991, est=3.084441668568253, details={'actual_k': 200, 'was_impossible': False}),
 Prediction(uid='405', iid='300', r_ui=3.52990990990991, est=3.6820660088850325, details={'actual_k': 200, 'was_impossible': False}),
 Prediction(uid='405', iid='1244', r_ui=3.52990990990991, est=2.8471448281000766, details={'actual_k': 34, 'was_impossible': False})]

[Prediction(uid='405', iid='150', r_ui=3.52990990990991, est=3.889823903524877, details={'actual_k': 146, 'was_impossible': False}),

 Prediction(uid='405', iid='124', r_ui=3.52990990990991, est=3.917481332041734, details={'actual_k': 172, 'was_impossible': False}),

 Prediction(uid='405', iid='748', r_ui=3.52990990990991, est=3.084441668568253, details={'actual_k': 200, 'was_impossible': False}),

 Prediction(uid='405', iid='300', r_ui=3.52990990990991, est=3.6820660088850325, details={'actual_k': 200, 'was_impossible': False}),

 Prediction(uid='405', iid='1244', r_ui=3.52990990990991, est=2.8471448281000766, details={'actual_k': 34, 'was_impossible': False})]

відсортувати по спаданню, порекомендувати топ 20, 50 або...

In [ ]:

top_n = defaultdict(list)
for uid, iid, r_ui, est,details in predictions:
  top_n[uid].append((iid, est))

In [ ]:
top_n["405"][:5]

[('150', 3.889823903524877),
 ('124', 3.917481332041734),
 ('748', 3.084441668568253),
 ('300', 3.6820660088850325),
 ('1244', 2.8471448281000766)]

In [ ]:
for uid, user_ratings in top_n.items():
  top_n[uid] = sorted(user_ratings, key=lambda x: -x[1])


вивести перші рекомендації для юзерів

In [ ]:
for uid, user_ratings in list(top_n.items())[:10]:
  print(f"{uid=}", [x[0] for x in user_ratings[:10]])

uid='405' ['1189', '1599', '1467', '1293', '1201', '814', '1536', '1653', '1500', '1122']
uid='460' ['1189', '1599', '1467', '1293', '1201', '814', '1536', '1653', '1500', '1122']
uid='285' ['1189', '1599', '1467', '1293', '1201', '814', '1536', '1653', '1500', '1122']
uid='542' ['1189', '1599', '1467', '1293', '1201', '814', '1536', '1653', '1500', '1122']
uid='463' ['1189', '1599', '1467', '1293', '1201', '814', '1536', '1653', '1500', '1122']
uid='178' ['1189', '1599', '1467', '1293', '1201', '814', '1536', '1653', '1500', '1122']
uid='487' ['1189', '1599', '1467', '1293', '1201', '814', '1536', '1653', '1122', '1500']
uid='262' ['1189', '1599', '1467', '1201', '814', '1536', '1653', '1500', '1122', '1293']
uid='767' ['1189', '1599', '1467', '1293', '1201', '814', '1536', '1653', '1500', '1122']
uid='82' ['1189', '1599', '1467', '1293', '1201', '814', '1536', '1653', '1500', '1122']
